exchange-v2/contracts/ExchangeV2Core.sol

```diff
@@ -33,6 +34,65 @@ abstract contract ExchangeV2Core is Initializable, OwnableUpgradeable, AssetMatc
         emit Cancel(orderKeyHash);
     }
 
+    /**
+     * @dev function, generate sellOrder and buyOrder from parameters and call validateAndMatch() for purchase transaction
+     * @param direct struct with parameters for buy operation
+     * @param sellData data type V3_SELLV for left order
+     * @param buyData data type 3_BUY for right order
+     */
+    function directPurchase(
+        LibDirectPurchase.Purchase memory direct,
+        bytes memory sellData,
+        bytes memory buyData
+    ) external payable {
+        bytes memory nftAssetData = abi.encode(direct.token, direct.tokenId);
+        LibAsset.Asset memory nft = LibAsset.Asset(LibAsset.AssetType(direct.assetType, nftAssetData), direct.tokenAmount);
+        LibAsset.Asset memory payment = LibAsset.Asset(LibAsset.AssetType(LibAsset.ETH_ASSET_CLASS, ""), direct.price);
+
+        LibOrder.Order memory orderLeft = LibOrder.Order(direct.seller, nft, address(0), payment, direct.salt, 0, 0, LibOrderDataV3.V3_SELL, sellData);
+        LibOrder.Order memory orderRight = LibOrder.Order(msg.sender, payment, address(0), nft, 0, 0, 0, LibOrderDataV3.V3_BUY, buyData);
+        validateAndMatch(orderLeft, direct.signature, orderRight);
+    }
+
+    /**
+     * @dev function, generate sellOrder and buyOrder from parameters and call validateAndMatch() for accept bid transaction
+     * @param direct struct with parameters for accept bid operation
+     * @param sellData data type V3_BUY for left order
+     * @param buyData data type V3_SELL for left order
+     */
+    function directAcceptBid(
+        LibDirectPurchase.AcceptBid memory direct,
+        bytes memory buyData,
+        bytes memory sellData
+    ) external payable {
+        bytes memory paymentAssetData = abi.encode(direct.tokenPayment);
+        bytes memory nftAssetData = abi.encode(direct.tokenNft, direct.tokenId);
+        LibAsset.Asset memory payment = LibAsset.Asset(LibAsset.AssetType(LibAsset.ERC20_ASSET_CLASS, paymentAssetData), direct.price);
+        LibAsset.Asset memory nft = LibAsset.Asset(LibAsset.AssetType(direct.assetType, nftAssetData), direct.tokenAmount);
+
+        LibOrder.Order memory orderLeft = LibOrder.Order(direct.buyer, payment, address(0), nft, direct.salt, 0, 0, LibOrderDataV3.V3_BUY, buyData);
+        LibOrder.Order memory orderRight = LibOrder.Order(msg.sender, nft, address(0), payment, 0, 0, 0, LibOrderDataV3.V3_SELL, sellData);
+        validateAndMatch(orderLeft, direct.signature, orderRight);
+    }
+
+    /**
+      * @dev function, validate orders and call matchAndTransfer()
+      * @param orderLeft left order
+      * @param signatureLeft order left signature
+      * @param orderRight right order
+      */
+    function validateAndMatch(LibOrder.Order memory orderLeft, bytes memory signatureLeft, LibOrder.Order memory orderRight) internal {
```

Comment:
this could probably be more generic
for example it could take second signature, so we can this method in matchOrders() also

Reply:
done